packages/opentelemetry-core/src/context/propagation/HttpTraceContext.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  SpanContext,
+  HttpTextFormat,
+  TraceOptions,
+} from '@opentelemetry/types';
+import { TraceState } from '../../trace/TraceState';
+
+export const TRACE_PARENT_HEADER = 'traceparent';
+export const TRACE_STATE_HEADER = 'tracestate';
+const VALID_TRACE_PARENT_REGEX = /^[\da-f]{2}-[\da-f]{32}-[\da-f]{16}-[\da-f]{2}$/;
+const VALID_TRACEID_REGEX = /^[0-9a-f]{32}$/i;
+const VALID_SPANID_REGEX = /^[0-9a-f]{16}$/i;
+const INVALID_ID_REGEX = /^0+$/i;
+const VERSION = '00';
+
+function parse(traceParent: string): SpanContext | null {
+  const match = traceParent.match(VALID_TRACE_PARENT_REGEX);
+  if (!match) return null;
+  const parts = traceParent.split('-');
+  const [version, traceId, spanId, option] = parts;
+  // tslint:disable-next-line:ban Needed to parse hexadecimal.
+  const traceOptions = parseInt(option, 16);
+
+  if (
+    !isValidVersion(version) ||
+    !isValidTraceId(traceId) ||
+    !isValidSpanId(spanId)
+  ) {
+    return null;
+  }
+
+  return { traceId, spanId, traceOptions };
+}
+
+function isValidVersion(version: string): boolean {
+  return version === VERSION;
+}
+
+function isValidTraceId(traceId: string): boolean {
+  return VALID_TRACEID_REGEX.test(traceId) && !INVALID_ID_REGEX.test(traceId);
+}
+
+function isValidSpanId(spanId: string): boolean {
+  return VALID_SPANID_REGEX.test(spanId) && !INVALID_ID_REGEX.test(spanId);
+}
+
+/**
+ * Propagates {@link SpanContext} through Trace Context format propagation.
+ *
+ * Based on the Trace Context specification:
+ * https://www.w3.org/TR/trace-context/
+ */
+export class HttpTraceContext implements HttpTextFormat {
+  inject(
+    spanContext: SpanContext,
+    format: string,
+    carrier: { [key: string]: unknown }
```

Comment:
It's kind of cool that `carrier` here can have a more specific type to store the actual headers. It also makes it clear that the header injection happens syncronously, whereas the callback-based design could make you wonder if it could be asyncronous. 

However, in the browser we would actually likely be setting the headers via [XMLHttpRequest.setRequestHeader](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/setRequestHeader), which means that we would need to have a temporary object to receive the headers and then read them from that object.

How was it decided to make `carrier` an object vs. the previous `HeaderSetter`/`HeaderGetter` approach?

Reply:
This originating from https://github.com/open-telemetry/opentelemetry-js/pull/70#discussion_r299725331, @rochdev Could you please elaborate more on this?

Reply:
My point was that the carrier should be any arbitrary `object` type, which includes but is not limited to object literals and `XMLHttpRequest`. The idea is to be flexible in the implementation, which is necessary as we can see from the comment above. In general the `inject` and `extract` methods should be as flexible as possible to allow extensions.

Reply:
Yes, I agree - couldn't we accomplish that by passing in a `HeaderSetter` and `HeaderGetter`? Then you basically just given in a function that translates to whatever format/object you want. Or am I missing something here?

Reply:
I'm not a fan of this since it's overly complex for simple objects. Should there simply be a different implementation for Node and for the browser? It also looks like a case where TextMap and HTTP are not the same since I would expect the browser to also support a simple object for TextMap. This is yet another area where the spec falls short.

Reply:
OK, I think for the browser we can just pass in an object and then get the headers out of it, which has benefits to its design, so I'm fine with this approach overall.

We will definitely need a different instrumentation package for XHR in the browser vs. http for Node, so perhaps that's where the different implementations could live.